eni: Fix data race during multiple updates at once

christarazi opened the pull request:
This data race was found by modifying the
(*ENISuite).TestNodeManagerDefaultAllocation test to inject spurious
update calls to the IPAM NodeManager for the CiliumNode that's already
being updated.

Fixes:

```
WARNING: DATA RACE
Read at 0x00c0005a09e8 by goroutine 36:
  github.com/cilium/cilium/pkg/aws/eni.(*Node).CreateInterface()
      /home/chris/code/cilium/cilium/pkg/aws/eni/node.go:358 +0x4aa
  github.com/cilium/cilium/pkg/ipam.(*Node).createInterface()
      /home/chris/code/cilium/cilium/pkg/ipam/node.go:441 +0x2a1
  github.com/cilium/cilium/pkg/ipam.(*Node).maintainIPPool()
      /home/chris/code/cilium/cilium/pkg/ipam/node.go:634 +0x868
  github.com/cilium/cilium/pkg/ipam.(*Node).MaintainIPPool()
      /home/chris/code/cilium/cilium/pkg/ipam/node.go:678 +0xd0
  github.com/cilium/cilium/pkg/ipam.(*NodeManager).Update.func2()
      /home/chris/code/cilium/cilium/pkg/ipam/node_manager.go:279 +0xab
  github.com/cilium/cilium/pkg/trigger.(*Trigger).waiter()
      /home/chris/code/cilium/cilium/pkg/trigger/trigger.go:206 +0x4db

Previous write at 0x00c0005a09e8 by goroutine 40:
  github.com/cilium/cilium/pkg/aws/eni.(*Node).UpdatedNode()
      /home/chris/code/cilium/cilium/pkg/aws/eni/node.go:66 +0x85
  github.com/cilium/cilium/pkg/ipam.(*Node).UpdatedResource()
      /home/chris/code/cilium/cilium/pkg/ipam/node.go:325 +0x81
  github.com/cilium/cilium/pkg/ipam.(*NodeManager).Update.func1()
      /home/chris/code/cilium/cilium/pkg/ipam/node_manager.go:263 +0x8e
  github.com/cilium/cilium/pkg/ipam.(*NodeManager).Update()
      /home/chris/code/cilium/cilium/pkg/ipam/node_manager.go:321 +0x1e6
  github.com/cilium/cilium/pkg/aws/eni.(*ENISuite).TestNodeManagerDefaultAllocation.func2()
      /home/chris/code/cilium/cilium/pkg/aws/eni/node_manager_test.go:276 +0x6f

Goroutine 36 (running) created at:
  github.com/cilium/cilium/pkg/trigger.NewTrigger()
      /home/chris/code/cilium/cilium/pkg/trigger/trigger.go:129 +0x24e
  github.com/cilium/cilium/pkg/ipam.(*NodeManager).Update()
      /home/chris/code/cilium/cilium/pkg/ipam/node_manager.go:274 +0x659
  github.com/cilium/cilium/pkg/aws/eni.(*ENISuite).TestNodeManagerDefaultAllocation()
      /home/chris/code/cilium/cilium/pkg/aws/eni/node_manager_test.go:262 +0xc15
  runtime.call32()
      /usr/lib/go/src/runtime/asm_amd64.s:540 +0x3d
  reflect.Value.Call()
      /usr/lib/go/src/reflect/value.go:336 +0xd8
  gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1()
      /home/chris/code/cilium/cilium/vendor/gopkg.in/check.v1/check.go:781 +0xabb
  gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1()
      /home/chris/code/cilium/cilium/vendor/gopkg.in/check.v1/check.go:675 +0xe1

Goroutine 40 (running) created at:
  github.com/cilium/cilium/pkg/aws/eni.(*ENISuite).TestNodeManagerDefaultAllocation()
      /home/chris/code/cilium/cilium/pkg/aws/eni/node_manager_test.go:274 +0x10b2
  runtime.call32()
      /usr/lib/go/src/runtime/asm_amd64.s:540 +0x3d
  reflect.Value.Call()
      /usr/lib/go/src/reflect/value.go:336 +0xd8
  gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1()
      /home/chris/code/cilium/cilium/vendor/gopkg.in/check.v1/check.go:781 +0xabb
  gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1()
      /home/chris/code/cilium/cilium/vendor/gopkg.in/check.v1/check.go:675 +0xe1
```

Fixes: https://github.com/cilium/cilium/issues/13251

Validated fix by deploying connectivity check on EKS.

```release-note
Fix potential bug in ENI IPAM when multiple updates at the same time are performed to the a CiliumNode resource
```

christarazi committed:
eni: Fix data race during multiple updates at once

This data race was found by modifying the
(*ENISuite).TestNodeManagerDefaultAllocation test to inject spurious
update calls to the IPAM NodeManager for the CiliumNode that's already
being updated.

Fixes:

```
WARNING: DATA RACE
Read at 0x00c0005a09e8 by goroutine 36:
  github.com/cilium/cilium/pkg/aws/eni.(*Node).CreateInterface()
      /home/chris/code/cilium/cilium/pkg/aws/eni/node.go:358 +0x4aa
  github.com/cilium/cilium/pkg/ipam.(*Node).createInterface()
      /home/chris/code/cilium/cilium/pkg/ipam/node.go:441 +0x2a1
  github.com/cilium/cilium/pkg/ipam.(*Node).maintainIPPool()
      /home/chris/code/cilium/cilium/pkg/ipam/node.go:634 +0x868
  github.com/cilium/cilium/pkg/ipam.(*Node).MaintainIPPool()
      /home/chris/code/cilium/cilium/pkg/ipam/node.go:678 +0xd0
  github.com/cilium/cilium/pkg/ipam.(*NodeManager).Update.func2()
      /home/chris/code/cilium/cilium/pkg/ipam/node_manager.go:279 +0xab
  github.com/cilium/cilium/pkg/trigger.(*Trigger).waiter()
      /home/chris/code/cilium/cilium/pkg/trigger/trigger.go:206 +0x4db

Previous write at 0x00c0005a09e8 by goroutine 40:
  github.com/cilium/cilium/pkg/aws/eni.(*Node).UpdatedNode()
      /home/chris/code/cilium/cilium/pkg/aws/eni/node.go:66 +0x85
  github.com/cilium/cilium/pkg/ipam.(*Node).UpdatedResource()
      /home/chris/code/cilium/cilium/pkg/ipam/node.go:325 +0x81
  github.com/cilium/cilium/pkg/ipam.(*NodeManager).Update.func1()
      /home/chris/code/cilium/cilium/pkg/ipam/node_manager.go:263 +0x8e
  github.com/cilium/cilium/pkg/ipam.(*NodeManager).Update()
      /home/chris/code/cilium/cilium/pkg/ipam/node_manager.go:321 +0x1e6
  github.com/cilium/cilium/pkg/aws/eni.(*ENISuite).TestNodeManagerDefaultAllocation.func2()
      /home/chris/code/cilium/cilium/pkg/aws/eni/node_manager_test.go:276 +0x6f

Goroutine 36 (running) created at:
  github.com/cilium/cilium/pkg/trigger.NewTrigger()
      /home/chris/code/cilium/cilium/pkg/trigger/trigger.go:129 +0x24e
  github.com/cilium/cilium/pkg/ipam.(*NodeManager).Update()
      /home/chris/code/cilium/cilium/pkg/ipam/node_manager.go:274 +0x659
  github.com/cilium/cilium/pkg/aws/eni.(*ENISuite).TestNodeManagerDefaultAllocation()
      /home/chris/code/cilium/cilium/pkg/aws/eni/node_manager_test.go:262 +0xc15
  runtime.call32()
      /usr/lib/go/src/runtime/asm_amd64.s:540 +0x3d
  reflect.Value.Call()
      /usr/lib/go/src/reflect/value.go:336 +0xd8
  gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1()
      /home/chris/code/cilium/cilium/vendor/gopkg.in/check.v1/check.go:781 +0xabb
  gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1()
      /home/chris/code/cilium/cilium/vendor/gopkg.in/check.v1/check.go:675 +0xe1

Goroutine 40 (running) created at:
  github.com/cilium/cilium/pkg/aws/eni.(*ENISuite).TestNodeManagerDefaultAllocation()
      /home/chris/code/cilium/cilium/pkg/aws/eni/node_manager_test.go:274 +0x10b2
  runtime.call32()
      /usr/lib/go/src/runtime/asm_amd64.s:540 +0x3d
  reflect.Value.Call()
      /usr/lib/go/src/reflect/value.go:336 +0xd8
  gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1()
      /home/chris/code/cilium/cilium/vendor/gopkg.in/check.v1/check.go:781 +0xabb
  gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1()
      /home/chris/code/cilium/cilium/vendor/gopkg.in/check.v1/check.go:675 +0xe1
```

Signed-off-by: Chris Tarazi <chris@isovalent.com>

aanm commented on pkg/aws/eni/node.go:
```diff
@@ -336,26 +335,43 @@ func (n *Node) CreateInterface(ctx context.Context, allocation *ipam.AllocationA
-	desc := "Cilium-CNI (" + n.node.InstanceID() + ")"
+	n.mutex.RLock()
+	instanceID := n.node.InstanceID()
+	n.mutex.RUnlock()
```
I've double checked, the `n.node` does not need to be protected by the mutex. We can drop this change and also modify the location of `n.mutex` to be below `node *ipam.Node`, on lines 46-49, to avoid confusion.

christarazi committed:
eni: Signal to avoid locking IPAM node in ENI

This field does not need to be protected by the mutex because it is
already protected by its own mutex. This commit moves the declaration of
the mutex field to be below the ipam.Node field to indicate that it's
not protected by the mutex, and therefore does not make any functional
changes.

Signed-off-by: Chris Tarazi <chris@isovalent.com>
